File: frontend/src/engine/finance/fixedexpense.test.ts
import { describe, expect, it } from 'vitest'
import { Money } from '@/engine/decimal'
import { activeIn, resolveAsOf as resolveFixedAmount, sumAsOf } from '@/engine/finance/fixedexpense'
import { addMonths } from '@/engine/finance/period'
import type { FixedExpenseAmountRow } from '@/engine/finance/models'

const amounts: FixedExpenseAmountRow[] = [
  // Deliberately out of order: resolveAsOf must not assume sorting.
  { fixedExpenseId: 1, effectiveFrom: '2026-05', amount: '12000' },
  { fixedExpenseId: 1, effectiveFrom: '2026-01', amount: '10000' },
  { fixedExpenseId: 1, effectiveFrom: '2026-09', amount: '15000' },
]

describe('resolveFixedAmount', () => {
  it('toma el mayor effective_from <= period', () => {
    expect(resolveFixedAmount(amounts, '2026-04').toString()).toBe('10000')
    expect(resolveFixedAmount(amounts, '2026-05').toString()).toBe('12000')
    expect(resolveFixedAmount(amounts, '2026-08').toString()).toBe('12000')
    expect(resolveFixedAmount(amounts, '2027-01').toString()).toBe('15000')
  })
  it('borde exacto: effective_from == period aplica', () => {
    expect(resolveFixedAmount(amounts, '2026-09').toString()).toBe('15000')
  })
  it('sin entrada aplicable aún devuelve cero', () => {
    expect(resolveFixedAmount(amounts, '2025-12').toString()).toBe('0')
    expect(resolveFixedAmount([], '2026-01').toString()).toBe('0')
  })
})

// The original month-by-month walk, kept as the oracle sumAsOf must match
// (mirror of TestSumAsOfMatchesMonthByMonth in Go).
function sumMonthByMonth(from: string, to: string): Money {
  let total = Money.zero()
  for (let m = from; m <= to; m = addMonths(m, 1)) total = total.add(resolveFixedAmount(amounts, m))
  return total
}

describe('sumAsOf', () => {
  it.each([
    ['rango completo con tres tramos', '2026-01', '2027-03'],
    ['empieza antes del primer monto (meses en cero)', '2025-06', '2026-02'],
    ['un solo mes', '2026-05', '2026-05'],
    ['termina justo antes de un cambio', '2026-02', '2026-08'],
    ['rango vacío (from > to)', '2026-05', '2026-04'],
  ])('%s', (_name, from, to) => {
    expect(sumAsOf(amounts, from, to).toString()).toBe(sumMonthByMonth(from, to).toString())
  })
})

describe('activeIn', () => {
  it('antes del inicio no factura', () => {
    expect(activeIn({ startPeriod: '2026-03', endPeriod: '' }, '2026-02')).toBe(false)
  })
  it('activo para siempre cuando endPeriod es vacío', () => {
    expect(activeIn({ startPeriod: '2026-03', endPeriod: '' }, '2030-01')).toBe(true)
  })
  it('después del fin no factura; el mes final sí', () => {
    const fe = { startPeriod: '2026-03', endPeriod: '2026-06' }
    expect(activeIn(fe, '2026-06')).toBe(true)
    expect(activeIn(fe, '2026-07')).toBe(false)
  })
})

File: backend/finance/fixedexpense_test.go
package finance

import (
	"testing"

	"github.com/gastonlarap-a11y/app-finance/backend/shared/types"
)

func amt(t *testing.T, s string) types.Decimal {
	t.Helper()
	d, err := types.New(s)
	if err != nil {
		t.Fatalf("types.New(%q): %v", s, err)
	}
	return d
}

// sumMonthByMonth is the original O(months) implementation, kept as the oracle
// that sumAsOf must match.
func sumMonthByMonth(rows []FixedExpenseAmount, from, to string) types.Decimal {
	total := types.Zero()
	for m := from; m <= to; m = addMonths(m, 1) {
		total = total.Add(resolveAsOf(rows, m))
	}
	return total
}

func TestSumAsOfMatchesMonthByMonth(t *testing.T) {
	rows := []FixedExpenseAmount{
		// Deliberately unsorted: sumAsOf must not depend on input order.
		{EffectiveFrom: "2026-06", Amount: amt(t, "12000")},
		{EffectiveFrom: "2025-11", Amount: amt(t, "9990")},
		{EffectiveFrom: "2026-01", Amount: amt(t, "10500.50")},
	}
	tests := []struct {
		name     string
		from, to string
	}{
		{"rango completo con tres tramos", "2025-11", "2027-03"},
		{"empieza antes del primer monto (meses en cero)", "2025-06", "2026-02"},
		{"un solo mes", "2026-06", "2026-06"},
		{"termina justo antes de un cambio", "2025-12", "2026-05"},
		{"rango vacío (from > to)", "2026-05", "2026-04"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := sumAsOf(rows, tt.from, tt.to)
			want := sumMonthByMonth(rows, tt.from, tt.to)
			if got.Cmp(want) != 0 {
				t.Fatalf("sumAsOf(%s..%s) = %s, want %s", tt.from, tt.to, got, want)
			}
		})
	}
}

func TestLatestAsOf(t *testing.T) {
	rows := []FixedExpenseAmount{
		{EffectiveFrom: "2026-03", Amount: amt(t, "2")},
		{EffectiveFrom: "2026-01", Amount: amt(t, "1")},
	}
	tests := []struct {
		name     string
		period   string
		wantOK   bool
		wantFrom string
	}{
		{"antes de cualquier fila", "2025-12", false, ""},
		{"en la primera fila", "2026-01", true, "2026-01"},
		{"entre filas usa la anterior", "2026-02", true, "2026-01"},
		{"después de la última", "2027-01", true, "2026-03"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			row, ok := latestAsOf(rows, tt.period)
			if ok != tt.wantOK || (ok && row.EffectiveFrom != tt.wantFrom) {
				t.Fatalf("latestAsOf(%s) = (%+v, %v), want from %q ok %v", tt.period, row, ok, tt.wantFrom, tt.wantOK)
			}
		})
	}
}
